Targets transition mega PR (#2)
USGS-R/ds-pipelines-targets-2-course: responses/11-targets-debug.md

> Breakpoints allow you to run a function (or script) up until the line of the breakpoint, and then the evaluation pauses. You are able to inspect all variables available at that point in the evaluation, and even step carefully forward one line at a time. It is out of scope of this exercise to go through exactly how to use debuggers, but they are powerful and helpful tools. It would be a good idea to read up on them if you haven't run into breakpoints yet. 

Reviewer: Add a link to the RStudio debugging doc page: https://support.rstudio.com/hc/en-us/articles/205612627-Debugging-with-RStudio

In `targets`, you can't set a breakpoint in the "normal" way, which would be clicking on the line number after you sourced the script. Instead, you need to use the other method for debugging in R, which requires adding the function call `browser()` to the line where you'd like the function call to stop and specifying an additional argument when you call `tar_make()`. 

---

> You have a working, albeit brittle, pipeline in your course repository. You can try it out with `targets::tar_make()`. This pipeline has a number of things you'll work to fix later, but for now, it is a useful reference. The pipeline contains several functions which are defined in `.R` files. 

Reviewer: ```suggestion
You have a working, albeit brittle, pipeline in your course repository. You can try it out with `targets::tar_make()`. This pipeline has a number of things you'll work to fix later, but for now, it is a useful reference. The pipeline contains a _targets.R file and several functions defined in `.R` files. 
```

Reviewer: See comment in 22-target-refactor.md, which provides directions for how to clone the template repo - do we want them calling tar_make() already on this pipeline, and if so, should the cloning instructions be here instead of in step 22?

Author: Ah! I didn't even notice that. I suppose that was an issue with the previous course, as well. I think we should move the cloning instructions up. I wonder if it should be in an earlier step (perhaps the "overview" one?) so that it is set up way ahead of time and users can choose to follow along with the pipeline code on GitHub or locally.

So, if you wanted to look at what `download_files` where created within the `download_nwis_data` function, you could set a breakpoint by adding `browser()` to the `"1_fetch/src/get_nwis_data.R"` file. Hint: to quickly navigate to this function source code from your makefile, you can put your cursor on the name of the function then click F2 and it will take you to the correct location in the corresponding source file!

![browser()](https://user-images.githubusercontent.com/2349007/82158816-bed9bb00-984f-11ea-8892-b2aeb5e4818d.png)

> There is one more step to get your breakpoint to work in `targets`. You will need to add `callr_function = NULL` to your `tar_make()` call. When you run `tar_make(callr_function = NULL)`, you will land right in the middle of line 8. Give it a try on your own.

Reviewer: The **targets** documentation also says you need to start from a fresh R session before debugging. I have yet to see or understand why, and I also don't know whether it's useful/important to restart R _after_ debugging. Do you know? Should we also mention this?

Author: Hmmm I don't remember reading that and when I was trying this debugging stuff out, I definitely didn't restart R before or after my debugging tests. Let me do a few tests and see what happens.

Author: I am still unimpacted by this, but from searching through the documentation, it looks like this may be if you are setting `debug` in the `tar_option_set()` function: https://github.com/ropensci/targets/blob/7a76cf8c94145f2f4b70f0859dba7eb5d918396c/R/tar_option_set.R#L60-L61

Though, I am not sure when you need to do that vs adding `browser()` to a particular function and running `tar_make(callr_function = NULL)`.

Author: I am going to skip mentioning since I have yet to need to follow these guidelines.

> To navigate while in browser mode, you can use the buttons at the top of your console pane:

Reviewer: See https://github.com/USGS-R/ds-pipelines-targets-2-course/issues/3 for my Mac experience - the upshot is that I needed to update RStudio. It's still possible (though increasingly unlikely over time) that students of this course will need to do the same.

Author: Adding a short comment about RStudio versions in this section.

> ![debugnav](https://support.rstudio.com/hc/en-us/article_attachments/201594747/console-toolbar.png)

Reviewer: Consider adding a few suggestions for things to explore while debugging this example function, to give students an actual taste of debugging rather than letting them figure it out on their own many weeks/months later.

Reviewer: Also add a reminder to remove the `browser()` call when they're done.

> ---

Reviewer: Side note re: template code:
```r
data_out <- data.frame(agency_cd = c(), site_no = c(), dateTime = c(), 
                         X_00010_00000 = c(), X_00010_00000_cd = c(), tz_cd = c())
```
creates a data.frame with 0 rows (intended) but also 0 columns (not intended). The code still works because `rbind(NULL, new_df)` works, but it'd be just as easy and slightly more transparent to set
```r
data_out <- NULL
```

Also, probably better to model `bind_rows` rather than `rbind` in this function.

Author: Done! Except that I used `data.frame()` because I like to see the type of data structure that it will become in the setup part.

https://github.com/USGS-R/ds-pipelines-targets-2-template/commit/709c508c7865be15fbe7e9583c28c05b25a0a003
